### PR TITLE
Kafka Connect: add option to force columns to lowercase

### PR DESCRIPTION
**Pull request by yornstei**
The current connector will create the columns with the same casing as the incoming data, even though many catalogs will show the columns as lowercase. This PR will give the option to force columns to lowercase. 

Refer to this [slack thread](https://apache-iceberg.slack.com/archives/C03SEK9PW0Z/p1725468165932919) for more details.

cc @bryanck

### CODE REVIEW

**Review by ajantha-bhat — COMMENTED** [review]
Isn't this now a conflicting property with "iceberg.tables.schema-case-insensitive", what if case sensitive enabled and this property is also enabled? Isn't this wrong?


Also, please add the testcases for the new changes.

**Comment by bryanck** [comment]
Note, this addresses https://github.com/apache/iceberg/issues/11091

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java [hunk 1]
```diff
@@ -79,6 +79,8 @@ public class IcebergSinkConfig extends AbstractConfig {
       "iceberg.tables.schema-force-optional";
   private static final String TABLES_SCHEMA_CASE_INSENSITIVE_PROP =
       "iceberg.tables.schema-case-insensitive";
+  private static final String TABLES_SCHEMA_FORCE_COLUMNS_TO_LOWERCASE_PROP =
+      "iceberg.tables.schema-force-columns-to-lowercase";
```
Comment (bryanck) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:83]: I think `iceberg.tables.schema-force-lowercase` is sufficient and consistent with `schema-force-optional`. Also, moving this up one to be below `schema-force optional` would be preferable.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java [hunk 2]
```diff
@@ -162,6 +164,12 @@ private static ConfigDef newConfigDef() {
         false,
         Importance.MEDIUM,
         "Set to true to look up table columns by case-insensitive name, false for case-sensitive");
+    configDef.define(
+        TABLES_SCHEMA_FORCE_COLUMNS_TO_LOWERCASE_PROP,
+        ConfigDef.Type.BOOLEAN,
+        false,
+        Importance.MEDIUM,
+        "Set to true to force table column names to lowercase, false for preserve case");
```
Comment (bryanck) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:172]: Same note here on name and moving it up one element to be next to "force optional".

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java [hunk 3]
```diff
@@ -402,6 +416,10 @@ public boolean schemaCaseInsensitive() {
     return getBoolean(TABLES_SCHEMA_CASE_INSENSITIVE_PROP);
   }
 
+  public boolean schemaForceColumnsToLowercase() {
```
Comment (bryanck) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:419]: Better if we move this up below to "force optional".

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java [hunk 4]
```diff
@@ -333,10 +341,16 @@ public TableSinkConfig tableConfig(String tableName) {
           Pattern routeRegex = routeRegexStr == null ? null : Pattern.compile(routeRegexStr);
 
           String idColumnsStr = tableConfig.getOrDefault(ID_COLUMNS, tablesDefaultIdColumns());
+          if (schemaForceColumnsToLowercase() && idColumnsStr != null) {
+            idColumnsStr = idColumnsStr.toLowerCase();
```
Comment (bryanck) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:345]: I feel we should remove this, it is defined in the config so a user can set this to lower case as needed.

File: kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java [hunk 5]
```diff
@@ -333,10 +341,16 @@ public TableSinkConfig tableConfig(String tableName) {
...
+          }
           List<String> idColumns = stringToList(idColumnsStr, ",");
 
           String partitionByStr =
               tableConfig.getOrDefault(PARTITION_BY, tablesDefaultPartitionBy());
+          if (schemaForceColumnsToLowercase() && partitionByStr != null) {
+            partitionByStr = partitionByStr.toLowerCase();
```
Comment (bryanck) [inline kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:352]: Same, here, I don't think we need this.

**Comment by bryanck** [comment]
> Isn't this now a conflicting property with "iceberg.tables.schema-case-insensitive", what if case sensitive enabled and this property is also enabled? Isn't this wrong?
> 
> Also, please add the testcases for the new changes.

A test would be good to add. It is fine to keep two properties, one for reading from the schema, and one for creating the schema. If we combined them that would be a change in behavior for current users of the sink.

**Comment by bryanck** [comment]
We should apply this config to schema updates as well, specifically add column.

Also, on further thought, this is getting somewhat convoluted, because if you enable force lowercase during create, but don't enable case insensitive lookups, then the column mappings won't be correct during record conversion. So I agree with @ajantha-bhat that we should think about having one configuration.

**Pull request closed without merging**